### ISSUE TITLE
Graph Editor Architecture

### ISSUE THREAD

**Issue by teodosin** [teodosin/karta#47]
#### Tracking issue and summary for determining the proper architecture for the graph entity crate. The goal is to separate this into a standalone crate that can be used by any bevy project. 
---

- [x] Make ui reactive to the underlying data entities #65 
- [x] Group all relevant files and submodules under a single module #73 
- [ ] Move forces to bevy_overlay_graph 
- [ ] Separate data and view nodes into their own entities #67 
- [ ] Clarify and clean up dependencies (must not be dependent on Karta, Karta depending on BOG is ok)
- [ ] Component property panel (blocked by #66)
- [ ] ? 

- [ ] bevy_overlay_crate git submodule
- [ ] bevy_overlay_crate public, usable through github
- [ ] 0.1 on crates.io

#### Resources

- https://www.howtogeek.com/devops/how-to-turn-a-turn-a-directory-in-a-git-repository-into-a-submodule/
- https://doc.rust-lang.org/cargo/guide/

**Comment by teodosin** [comment]
Initial thoughts: 
#48 Seems like a good aspirational goal. If Karta is eventually meant to appeal to programmers or writers, it would make sense that the entire UI / App would have support for keyboard only. Traditional keyboard shortcuts with the modifier keys as well as a vim-mode for those who don't want to move their hands from the home row. This could be under a feature flag. 

Touch and stylus support would also be a feature. 

Also down the line: Vr controller and glove support.

**Comment by teodosin** [comment]
How would this work as a crate that could be added to any bevy project?

Visualise entities and their relationships as well as system schedules, resources, and events.

Karta would then be an app that merely uses the graph crate to visualise itself. It would just manage the database of node networks, spawn nodes as entities and the graph representation would be generated by the graph crate just like for any other bevy app.

I personally still need a much deeper understanding of Bevy in order to make this work. I'm not clear on the separation between Karta's features and those that should belong in the crate. There might also need to be ergonomic ways to define custom node types, context menu actions and such on a per-app basis. And all this needs to play well with Karta's own scene system.

Not in scope of the thesis, I don't think.

**Comment by teodosin** [comment]
https://github.com/teodosin/bevy_overlay_graph

In preparation.

**Comment by teodosin** [comment]
Another _important_ thing that needs to be iterated on is how to separate the inputs of the underlying Bevy app and the graph crate. The graph is an application by itself with navigation and menus and such, but it should not interfere with the user's app inputs. There needs to be some kind of toggle that is able to disable the user's inputs to their own app from within the crate. At least that would be most ergonomic, so the user wouldn't have to manually make sure to disable their app inputs when they want to navigate the overlay. 

So far this feels like the biggest hurdle in making this usable as a crate.

**Comment by teodosin** [comment]
Sunday project: https://github.com/teodosin/bevy_overlay_graph now exists. It works; it spawns nodes for each entity with the GraphEntity component and displays their Name component on the node label. There's a lot of work to do there to reach feature parity with Karta's current build but without all the coupling.

I'm working on that crate in a separate workspace and using a dummy Bevy project to test it. I think that's a better path than refactoring and creating the crate while still inside Karta. This way I can build the crate to support what Karta needs without breaking Karta. Then when there's feature parity, Karta can be refactored to use the crate's API.

**Comment by teodosin** [comment]
Cleaning up some issues. Most of the work has been done in now deprecated repos which have been merged into this monorepo.
